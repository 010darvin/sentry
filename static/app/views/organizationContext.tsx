import {createContext} from 'react';

import {type Organization} from 'sentry/types';

export const OrganizationContext = createContext<Organization | null>(null);
