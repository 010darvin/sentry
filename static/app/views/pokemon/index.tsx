import Feature from 'sentry/components/acl/feature';
import Alert from 'sentry/components/alert';
import {t} from 'sentry/locale';
import {PageContent} from 'sentry/styles/organization';
import {Organization} from 'sentry/types';
import withOrganization from 'sentry/utils/withOrganization';

type Props = {
  children: React.ReactChildren;
  organization: Organization;
};

function PokemonContainer({organization}: Props) {
  return (
    <Feature
      hookName="feature-disabled:profiling-page"
      features={[]}
      organization={organization}
      renderDisabled={() => (
        <PageContent>
          <Alert type="warning">{t("You don't have access to this feature")}</Alert>
        </PageContent>
      )}
    >
      Foo bar
    </Feature>
  );
}

export default withOrganization(PokemonContainer);
