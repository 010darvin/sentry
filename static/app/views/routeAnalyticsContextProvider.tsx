import React, {createContext} from 'react';
import type {RouteContextInterface} from 'react-router';

import HookStore from 'sentry/stores/hookStore';
import {Organization} from 'sentry/types';

const DEFAULT_CONTEXT = {
  setDisableRouteAnalytics: () => {},
  setRouteAnalyticsParams: () => {},
  setOrganization: () => {},
};

export const RouteAnalyticsContext = createContext<{
  setDisableRouteAnalytics: (disable: boolean) => void;
  setOrganization: (organization: Organization) => void;
  setRouteAnalyticsParams: (params: Record<string, any>) => void;
}>(DEFAULT_CONTEXT);

interface Props extends RouteContextInterface {
  children?: React.ReactNode;
}

export default function RouteAnalyticsContextProvider({children, ...props}: Props) {
  const useRouteActivatedHook = HookStore.get('react-hook:route-activated')[0];
  const {setDisableRouteAnalytics, setRouteAnalyticsParams, setOrganization} =
    useRouteActivatedHook?.(props) || DEFAULT_CONTEXT;

  return (
    <RouteAnalyticsContext.Provider
      value={{
        setDisableRouteAnalytics,
        setRouteAnalyticsParams,
        setOrganization,
      }}
    >
      {children}
    </RouteAnalyticsContext.Provider>
  );
}
