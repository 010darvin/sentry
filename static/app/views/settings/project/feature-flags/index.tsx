import {Fragment, useEffect, useState} from 'react';
import styled from '@emotion/styled';
import isEqual from 'lodash/isEqual';

import {
  addErrorMessage,
  addLoadingMessage,
  addSuccessMessage,
} from 'sentry/actionCreators/indicator';
import {ModalRenderProps, openModal} from 'sentry/actionCreators/modal';
import Button from 'sentry/components/button';
import EmptyStateWarning from 'sentry/components/emptyStateWarning';
import FeatureBadge from 'sentry/components/featureBadge';
import CompactSelect from 'sentry/components/forms/compactSelect';
import ExternalLink from 'sentry/components/links/externalLink';
import SearchBar from 'sentry/components/searchBar';
import SentryDocumentTitle from 'sentry/components/sentryDocumentTitle';
import {IconAdd} from 'sentry/icons/iconAdd';
import {t, tct} from 'sentry/locale';
import ProjectStore from 'sentry/stores/projectsStore';
import space from 'sentry/styles/space';
import {Project} from 'sentry/types';
import {defined} from 'sentry/utils';
import handleXhrErrorResponse from 'sentry/utils/handleXhrErrorResponse';
import useApi from 'sentry/utils/useApi';
import useOrganization from 'sentry/utils/useOrganization';
import usePrevious from 'sentry/utils/usePrevious';
import SettingsPageHeader from 'sentry/views/settings/components/settingsPageHeader';
import TextBlock from 'sentry/views/settings/components/text/textBlock';
import PermissionAlert from 'sentry/views/settings/organization/permissionAlert';

import {FlagModal} from './modals/flagModal';
import {SegmentModal} from './modals/segmentModal';
import {Card} from './card';
import {Promo} from './promo';

type Props = ModalRenderProps & {
  project: Project;
};

export default function ProjectFeatureFlags({project}: Props) {
  const api = useApi();
  const organization = useOrganization();

  const currentFlags = project.featureFlags;
  const hasAccess = organization.access.includes('project:write');
  const disabled = !hasAccess;
  const previousFlags = usePrevious(currentFlags);

  const [flags, setFlags] = useState(currentFlags ?? {});
  const [query, setQuery] = useState('');
  const [status, setStatus] = useState(null);

  const showPromo = Object.keys(flags).length === 0;

  useEffect(() => {
    if (!isEqual(currentFlags, previousFlags)) {
      setFlags(currentFlags ?? {});
    }
  }, [currentFlags, previousFlags]);

  function handleAddFlag() {
    openModal(modalProps => (
      <FlagModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
      />
    ));
  }

  function handleEditFlag(flagKey: string) {
    openModal(modalProps => (
      <FlagModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
        flagKey={flagKey}
      />
    ));
  }

  async function handleDeleteFlag(flagKey: string) {
    const newFeatureFlags = {...flags};
    delete newFeatureFlags[flagKey];

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );
      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully deleted feature flag'));
    } catch (error) {
      const message = t('Unable to delete feature flag');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  async function handleActivateToggle(flagKey: string) {
    const newFeatureFlags = {...flags};
    newFeatureFlags[flagKey].enabled = !newFeatureFlags[flagKey].enabled;

    addLoadingMessage();

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );

      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully updated the feature flag'));
    } catch (error) {
      const message = t('Unable to update the feature flag');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  function handleAddSegment(flagKey: string) {
    openModal(modalProps => (
      <SegmentModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
        flagKey={flagKey}
      />
    ));
  }

  function handleEditSegment(flagKey: string, index: number) {
    openModal(modalProps => (
      <SegmentModal
        {...modalProps}
        organization={organization}
        project={project}
        flags={flags}
        flagKey={flagKey}
        segmentIndex={index}
      />
    ));
  }

  async function handleDeleteSegment(flagKey: string, index: number) {
    const newEvaluations = [...flags[flagKey].evaluation];
    newEvaluations.splice(index, 1);

    const newFeatureFlags = {...flags};
    newFeatureFlags[flagKey].evaluation = newEvaluations;

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );

      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully deleted segment'));
    } catch (error) {
      const message = t('Unable to delete segment');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  async function handleSortSegments(flagKey: string, segmentIds: string[]) {
    const sortedSegments = segmentIds
      .map(segmentId =>
        flags[flagKey].evaluation.find(segment => String(segment.id) === segmentId)
      )
      .filter(defined);

    const newFeatureFlags = {...flags};
    newFeatureFlags[flagKey].evaluation = sortedSegments;

    setFlags(newFeatureFlags);

    try {
      const result = await api.requestPromise(
        `/projects/${organization.slug}/${project.slug}/`,
        {
          method: 'PUT',
          data: {featureFlags: newFeatureFlags},
        }
      );
      ProjectStore.onUpdateSuccess(result);
      addSuccessMessage(t('Successfully sorted segments'));
    } catch (error) {
      setFlags(previousFlags ?? []);
      const message = t('Unable to sort segments');
      handleXhrErrorResponse(message)(error);
      addErrorMessage(message);
    }
  }

  const filteredFlags = Object.keys(flags).filter(key => {
    return (
      key.toLowerCase().includes(query.toLowerCase()) &&
      (status === null || flags[key].enabled === status)
    );
  });

  return (
    <SentryDocumentTitle title={t('Feature Flags')}>
      <Fragment>
        <SettingsPageHeader
          title={
            <Fragment>
              {t('Feature Flags')} <FeatureBadge type="beta" />
            </Fragment>
          }
          action={
            !showPromo && (
              <Button
                title={
                  disabled ? t('You do not have permission to add flags') : undefined
                }
                priority="primary"
                size="sm"
                icon={<IconAdd size="xs" isCircled />}
                onClick={handleAddFlag}
                disabled={disabled}
              >
                {t('Add Flag')}
              </Button>
            )
          }
        />
        <TextBlock>
          {tct(
            'Feature flags allow you to configure your code into different flavors by dynamically toggling certain functionality on and off. Learn more about feature flags in our [link:documentation].',
            {
              link: <ExternalLink href="" />,
            }
          )}
        </TextBlock>

        <PermissionAlert access={['project:write']} />

        {showPromo ? (
          <Promo hasAccess={hasAccess} onGetStarted={handleAddFlag} />
        ) : (
          <Content>
            <Filters>
              <SearchBar
                size="md"
                onChange={setQuery}
                query={query}
                placeholder={t('Search by flag name')}
              />
              <CompactSelect
                defaultValue={status}
                options={[
                  {value: null, label: t('All')},
                  {value: true, label: t('Active')},
                  {value: false, label: t('Inactive')},
                ]}
                value={status}
                onChange={value => setStatus(value.value)}
                triggerProps={{prefix: t('Status'), style: {width: '100%'}}}
              />
            </Filters>
            {!filteredFlags.length ? (
              <EmptyStateWarning>
                <p>{t('Sorry, no feature flag match your filters')}</p>
              </EmptyStateWarning>
            ) : (
              <div>
                {filteredFlags.map(filteredFlag => (
                  <Card
                    key={filteredFlag}
                    flagKey={filteredFlag}
                    kind={flags[filteredFlag].kind}
                    enabled={flags[filteredFlag].enabled}
                    description={flags[filteredFlag].description}
                    segments={flags[filteredFlag].evaluation}
                    onDelete={() => handleDeleteFlag(filteredFlag)}
                    onEdit={() => handleEditFlag(filteredFlag)}
                    onActivateToggle={() => handleActivateToggle(filteredFlag)}
                    onAddSegment={() => handleAddSegment(filteredFlag)}
                    onDeleteSegment={id => handleDeleteSegment(filteredFlag, id)}
                    onEditSegment={id => handleEditSegment(filteredFlag, id)}
                    hasAccess={hasAccess}
                    onSortSegments={({reorderedItems}) =>
                      handleSortSegments(filteredFlag, reorderedItems)
                    }
                  />
                ))}
              </div>
            )}
          </Content>
        )}
      </Fragment>
    </SentryDocumentTitle>
  );
}

const Content = styled('div')`
  display: grid;
  gap: ${space(2)};
`;

const Filters = styled('div')`
  display: grid;
  gap: ${space(1)};
  grid-template-columns: 1fr max-content;
`;
