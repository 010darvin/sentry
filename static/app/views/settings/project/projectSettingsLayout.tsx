import {cloneElement, isValidElement, useMemo} from 'react';
import {RouteComponentProps} from 'react-router';

import * as AppStoreConnectContext from 'sentry/components/projects/appStoreConnectContext';
import {Organization, Project} from 'sentry/types';
import useRouteAnalyticsParams from 'sentry/utils/useRouteAnalyticsParams';
import withOrganization from 'sentry/utils/withOrganization';
import ProjectContext from 'sentry/views/projects/projectContext';
import SettingsLayout from 'sentry/views/settings/components/settingsLayout';
import ProjectSettingsNavigation from 'sentry/views/settings/project/projectSettingsNavigation';

type Props = {
  children: React.ReactNode;
  organization: Organization;
} & RouteComponentProps<{orgId: string; projectId: string}, {}>;

type InnerProps = Props & {project: Project};

function InnerProjectSettingsLayout({
  params,
  routes,
  project,
  organization,
  children,
  ...props
}: InnerProps) {
  // set analytics params for route based analytics
  useRouteAnalyticsParams(
    // use memoized value to avoid re-rendering
    useMemo(
      () => ({
        project_id: project.id,
        project_platform: project.platform,
      }),
      [project.id, project.platform]
    )
  );
  return (
    <AppStoreConnectContext.Provider project={project} organization={organization}>
      <SettingsLayout
        params={params}
        routes={routes}
        {...props}
        renderNavigation={() => <ProjectSettingsNavigation organization={organization} />}
      >
        {children && isValidElement(children)
          ? cloneElement<any>(children, {organization, project})
          : children}
      </SettingsLayout>
    </AppStoreConnectContext.Provider>
  );
}

function ProjectSettingsLayout({params, ...props}: Props) {
  const {orgId, projectId} = params;

  return (
    <ProjectContext orgId={orgId} projectId={projectId}>
      {({project}) => <InnerProjectSettingsLayout {...{params, project, ...props}} />}
    </ProjectContext>
  );
}

export default withOrganization(ProjectSettingsLayout);
