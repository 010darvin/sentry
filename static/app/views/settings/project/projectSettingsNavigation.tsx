import {useContext} from 'react';

import AppStoreConnectContext from 'sentry/components/projects/appStoreConnectContext';
import {Organization, Project} from 'sentry/types';
import withProject from 'sentry/utils/withProject';
import SettingsNavigation from 'sentry/views/settings/components/settingsNavigation';
import getConfiguration from 'sentry/views/settings/project/navigationConfiguration';

type Props = {
  organization: Organization;
  project?: Project;
};

const ProjectSettingsNavigation = ({organization, project}: Props) => {
  const appStoreConnectContext = useContext(AppStoreConnectContext);

  const debugFilesNeedsReview = appStoreConnectContext
    ? Object.keys(appStoreConnectContext).some(
        // @ts-expect-error TS(2532) FIXME: Object is possibly 'undefined'.
        key => appStoreConnectContext[key].credentials.status === 'invalid'
      )
    : false;

  return (
    <SettingsNavigation
      navigationObjects={getConfiguration({project, organization, debugFilesNeedsReview})}
      access={new Set(organization.access)}
      features={new Set(organization.features)}
      organization={organization}
      project={project}
    />
  );
};

export default withProject(ProjectSettingsNavigation);
