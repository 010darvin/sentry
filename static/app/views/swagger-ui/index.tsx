import 'swagger-ui-react/swagger-ui.css';

import SwaggerUI from 'swagger-ui-react';

const swagger = require('../../../../tests/apidocs/openapi-derefed.json');

const SwaggerUIDocs = () => <SwaggerUI spec={swagger} />;

export default SwaggerUIDocs;
