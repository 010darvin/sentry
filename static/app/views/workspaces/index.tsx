import styled from '@emotion/styled';

import AsyncComponent from 'sentry/components/asyncComponent';
import PageHeading from 'sentry/components/pageHeading';
import {Panel} from 'sentry/components/panels';
import ConfigStore from 'sentry/stores/configStore';
import {PageContent} from 'sentry/styles/organization';
import space from 'sentry/styles/space';
import {Organization} from 'sentry/types';
import withOrganization from 'sentry/utils/withOrganization';

import WorkspaceCard from './components/workspaceCard';
import WorkspaceRow from './components/workspaceRow';
import {Workspace, workspaces} from './types';

type WorkspacesIndexProps = AsyncComponent['props'] & {
  organization: Organization;
};

type WorkspacesIndexState = AsyncComponent['state'] & {};

class WorkspacesIndex extends AsyncComponent<WorkspacesIndexProps, WorkspacesIndexState> {
  getEndpoints(): ReturnType<AsyncComponent['getEndpoints']> {
    const {organization} = this.props;
    return [['userTeams', `/organizations/${organization.slug}/user-teams/`]];
  }

  renderBody() {
    const user = ConfigStore.get('user');
    const {organization: _organization} = this.props;
    const {userTeams} = this.state;
    const workspaceMap: {
      all: Workspace[];
      member: Workspace[];
    } = workspaces.reduce(
      (map, workspace) => {
        const isTeamMember =
          workspace?.members[0].type === 'team' &&
          userTeams.some(team => team.id === workspace.members[0].id);
        const isUserMember =
          workspace?.members[0].type === 'user' && workspace.members[0].id === user.id;
        return isTeamMember || isUserMember
          ? {...map, member: [...map.member, workspace]}
          : {...map, all: [...map.all, workspace]};
      },
      {member: [] as Workspace[], all: [] as Workspace[]}
    );
    return (
      <PageContent>
        <PageHeading withMargins>Your Workspaces</PageHeading>
        <CardContainer>
          {workspaceMap.member.map(workspace => (
            <WorkspaceCard key={workspace.id} workspace={workspace} />
          ))}
        </CardContainer>
        <PageHeading withMargins>Other Workspaces</PageHeading>
        <Panel>
          {workspaceMap.all.map(workspace => (
            <WorkspaceRow key={workspace.id} workspace={workspace} />
          ))}
        </Panel>
      </PageContent>
    );
  }
}

const CardContainer = styled('div')`
  display: flex;
  gap: ${space(2)};
  margin-bottom: ${space(2)};
  align-items: flex-start;
`;

export default withOrganization(WorkspacesIndex);
