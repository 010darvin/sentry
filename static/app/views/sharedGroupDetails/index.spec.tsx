import {Event} from 'fixtures/js-stubs/event.js';
import {EventEntry} from 'fixtures/js-stubs/eventEntry.js';
import {EventStacktraceException} from 'fixtures/js-stubs/eventStacktraceException.js';
import {Group} from 'fixtures/js-stubs/group.js';
import {Project} from 'fixtures/js-stubs/project.js';
import {router} from 'fixtures/js-stubs/router.js';

import {render} from 'sentry-test/reactTestingLibrary';

import {RouteContext} from 'sentry/views/routeContext';
import SharedGroupDetails from 'sentry/views/sharedGroupDetails';

describe('SharedGroupDetails', function () {
  const eventEntry = EventEntry();
  const exception = EventStacktraceException().entries[0];
  const params = {shareId: 'a'};
  const router = router({params});

  beforeEach(function () {
    MockApiClient.addMockResponse({
      url: '/shared/issues/a/',
      body: Group({
        title: 'ZeroDivisionError',
        latestEvent: Event({
          entries: [eventEntry, exception],
        }),
        project: Project({organization: {slug: 'test-org'}}),
      }),
    });
  });

  afterEach(function () {
    MockApiClient.clearMockResponses();
  });

  it('renders', function () {
    const {container} = render(
      <RouteContext.Provider value={{router, ...router}}>
        <SharedGroupDetails
          params={params}
          api={new MockApiClient()}
          route={{}}
          router={router}
          routes={router.routes}
          routeParams={router.params}
          location={router.location}
        />
      </RouteContext.Provider>
    );

    expect(container).toSnapshot();
  });
});
