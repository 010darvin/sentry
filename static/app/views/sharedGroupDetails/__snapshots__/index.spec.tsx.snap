// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharedGroupDetails renders 1`] = `
<div>
  <div
    class="loading"
    data-test-id="loading-indicator"
  >
    <div
      class="loading-indicator"
    />
    <div
      class="loading-message"
    />
  </div>
</div>
`;
